File: examples/06-kubernetes/python/main.py
import os
import sys

import kflow
from kflow import Workflow, step, run
from handlers import (
    ingest_data, transform_data, validate_data,
    export_data, notify_complete, handle_error,
)


def main():
    wf = Workflow("data-pipeline")

    @wf.task("IngestData")
    def _ingest(input):
        return ingest_data(input)

    @wf.task("TransformData")
    def _transform(input):
        return transform_data(input)

    @wf.task("ValidateData")
    def _validate(input):
        return validate_data(input)

    @wf.task("ExportData")
    def _export(input):
        return export_data(input)

    @wf.task("NotifyComplete")
    def _notify(input):
        return notify_complete(input)

    @wf.task("HandleError")
    def _handle(input):
        return handle_error(input)

    wf.flow(
        step("IngestData").next("TransformData"),
        step("TransformData").next("ValidateData").retry(max_attempts=3, backoff_seconds=2),
        step("ValidateData").next("ExportData").catch("HandleError"),
        step("ExportData").next("NotifyComplete"),
        step("NotifyComplete").end(),
        step("HandleError").end(),
    )

    run(wf, input={"source": "s3://input/data.csv"})


if __name__ == "__main__":
    main()

File: examples/06-kubernetes/main.go
package main

import (
	"log"

	"github.com/pastorenue/kflow/examples/06-kubernetes/handlers"
	"github.com/pastorenue/kflow/pkg/kflow"
)

func main() {
	wf := kflow.New("data-pipeline")

	wf.Task("IngestData", handlers.IngestData)
	wf.Task("TransformData", handlers.TransformData).
		Retry(kflow.RetryPolicy{MaxAttempts: 3, BackoffSeconds: 2})
	wf.Task("ValidateData", handlers.ValidateData).Catch("HandleError")
	wf.Task("ExportData", handlers.ExportData)
	wf.Task("NotifyComplete", handlers.NotifyComplete)
	wf.Task("HandleError", handlers.HandleError)

	wf.Flow(
		kflow.Step("IngestData").Next("TransformData"),
		kflow.Step("TransformData").Next("ValidateData"),
		kflow.Step("ValidateData").Next("ExportData").Catch("HandleError"),
		kflow.Step("ExportData").Next("NotifyComplete"),
		kflow.Step("NotifyComplete").End(),
		kflow.Step("HandleError").End(),
	)

	if err := kflow.Dispatch(wf, kflow.Input{"source": "s3://input/data.csv"}); err != nil {
		log.Fatal(err)
	}
}
